### PR TITLE
Network integrated

### PR DESCRIPTION
**Pull request by benoit-robotics**
fix the incompatibility between depth and label images

### CODE REVIEW

File: scripts/segmentation_node [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python3
+#!/home/yemika/Public/anaconda3/envs/tensorflow_env/bin/python
```
Comment (benoit-robotics) [inline scripts/segmentation_node:1]: same problem as in the "cropping" branch, this cannot work

File: scripts/segmentation_node [hunk 2]
```diff
@@ -92,29 +130,46 @@ class SegmentationNode:
...
-        self.G = tf.get_default_graph()
-        self.x = self.G.get_tensor_by_name(self.network_config.image_input_tensor)
+        self.model = sm.Unet(self.backbone, encoder_weights='imagenet', classes=21, encoder_freeze=True)
+        self.model.load_weights(self.network_config.path)
         
-        self.dist = self.G.get_tensor_by_name(self.network_config.distribution_output_tensor)
-        tf.global_variables_initializer().run(session=self.sess)
-        print("Tf init finish")
+        # Predict some random image to load the necessary libraries
+        x = cv2.imread("/home/yemika/Mikael/Work/ODRS/SegMap/network/MobileNetSeg/data/train/images/image4914.png", cv2.IMREAD_COLOR)          # read the image as BGR
```
Comment (benoit-robotics) [inline scripts/segmentation_node:165]: same problem as in the "cropping" branch, this cannot work. This file doesn't exist
Remove this whole block it's useless ( line 165 - 172 )

**Comment by benoit-robotics** [comment]
@MikaYeghi You lost access to Slack, so please add your comments here

File: src/labeled_pc_map_node.cpp [hunk 3]
```diff
@@ -14,13 +14,14 @@
 #include <iostream>
 
 // for nclt: 14 classes
-POINT_CLOUD_REGISTER_POINT_STRUCT(pcl::PointSegmentedDistribution<14>,
+POINT_CLOUD_REGISTER_POINT_STRUCT(pcl::PointSegmentedDistribution<21>,
```
Comment (benoit-robotics) [inline src/labeled_pc_map_node.cpp:17]: you changed the number of classes from 14 to 21, does it mean that this code cannot work with the old pretrained network that only have 14 classes

Comment (MikaYeghi) [reply src/labeled_pc_map_node.cpp:17]: Yes. I thought it would make sense to change the hard-coded number to the variable storing the number of classes, but forgot to implement it. The authors hard-coded a lot of numbers, and it's a bit annoying.

File: scripts/segmentation_node (hunk 1 above)
Comment (MikaYeghi) [reply scripts/segmentation_node:1]: See my comments in the launchfile_description file. Need to set up the environment before running the package.

File: scripts/segmentation_node (hunk 2 above)
Comment (MikaYeghi) [reply scripts/segmentation_node:165]: You can remove it, I think I explained at some point why I did it, and it's not really mandatory.

File: scripts/segmentation_node (hunk 1 above)
Comment (benoit-robotics) [reply scripts/segmentation_node:1]: you should put the instructions in a txt file not a docx

Comment (benoit-robotics) [reply scripts/segmentation_node:1]: and anyway even if I follow your instructions it will never work, because there is your name in the path.

File: scripts/segmentation_node (hunk 2 above)
Comment (benoit-robotics) [reply scripts/segmentation_node:165]: I remember that, but you cannot commit an absolute path that has your name in it, it will never work on other computers

Comment (MikaYeghi) [reply scripts/segmentation_node:165]: I see. Will take it into account

File: scripts/segmentation_node (hunk 1 above)
Comment (MikaYeghi) [reply scripts/segmentation_node:1]: How do you suggest I reform this? Should I include a python environment in the package?

File: src/labeled_pc_map_node.cpp (hunk 3 above)
Comment (benoit-robotics) [reply src/labeled_pc_map_node.cpp:17]: I find that a little bit annoying that we cannot reused the old weights ( without changing this number back to 14 I mean )

Comment (MikaYeghi) [reply src/labeled_pc_map_node.cpp:17]: It's possible, but I'll have to make changes to their code, to remove these numbers, and it might take some time

File: scripts/segmentation_node (hunk 1 above)
Comment (benoit-robotics) [reply scripts/segmentation_node:1]: I will create a conda virtual env and see how I can modify the shebang

Comment (benoit-robotics) [reply scripts/segmentation_node:1]: Is conda installed on xolotl ? I get "conda: command not found"

Comment (MikaYeghi) [reply scripts/segmentation_node:1]: I guess you need to install it on your xolotl. It is installed on mine. See this link - https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html (anaconda installer)

Comment (MikaYeghi) [reply scripts/segmentation_node:1]: You can run `wget https://repo.anaconda.com/archive/Anaconda3-2021.05-Linux-x86_64.sh` in the terminal for downloading the .sh file, then simply `bash Anaconda-latest-Linux-x86_64.sh`.

Comment (benoit-robotics) [reply scripts/segmentation_node:1]: Ok thank you. I know that we already talk about that but your account on Slack has been deactivated so I don't have access to our conversation anymore. Remind me why "#!/usr/bin/python3" didn't work for you ?

Comment (benoit-robotics) [reply scripts/segmentation_node:1]: **#!/usr/bin/env python** might work

Comment (benoit-robotics) [reply scripts/segmentation_node:1]: or **#!/usr/bin/env python3** might work

File: src/labeled_pc_map_node.cpp (hunk 3 above)
Comment (benoit-robotics) [reply src/labeled_pc_map_node.cpp:17]: How would you do that ?

File: scripts/segmentation_node (hunk 1 above)
Comment (MikaYeghi) [reply scripts/segmentation_node:1]: The reason why it didn't work is the same as when you were trying to run the network training on the GPU - tensorflow wasn't installed properly (since it needs some side packages), so installing a special environment through conda can do the job. In any case, this will require certain packages to be installed in order to run the prediction step in the segmentation_node.

File: src/labeled_pc_map_node.cpp (hunk 3 above)
Comment (MikaYeghi) [reply src/labeled_pc_map_node.cpp:17]: I would try to do something like passing parameters from the launch file to the .cpp files, and replacing hard-coded numbers with variables. Something like this:
`int NUM_OF_CLASSES;
NUM_OF_CLASSES = {somehow reading from launch file...};
POINT_CLOUD_REGISTER_POINT_STRUCT(pcl::PointSegmentedDistribution<NUM_OF_CLASSES>`

Comment (benoit-robotics) [reply src/labeled_pc_map_node.cpp:17]: That won't work. The value of a template argument has to be known at compile time, but with what you are planning to do it's only known at runtime.

Comment (MikaYeghi) [reply src/labeled_pc_map_node.cpp:17]: I see. Then I don't fully understand why they made the class templated on the number of classes, and put that parameter in the launchfile. I suppose what I said would require to change the code a little, I don't know if it's that necessary...
